File: backend/adapters/docx_adapter.py
import io
from docx import Document
from docx.shared import Pt, Cm, RGBColor
from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.oxml.ns import qn
from docx.oxml import OxmlElement

PRIMARY   = RGBColor(0x4F, 0x46, 0xE5)  # indigo
MUTED     = RGBColor(0x64, 0x74, 0x8B)  # slate-500
BLACK     = RGBColor(0x1E, 0x29, 0x3B)  # slate-900

def _set_font(run, name="Arial", size=10, bold=False, color=None, italic=False):
    run.font.name = name
    run.font.size = Pt(size)
    run.font.bold = bold
    run.font.italic = italic
    if color:
        run.font.color.rgb = color

def _add_border_bottom(paragraph):
    pPr = paragraph._p.get_or_add_pPr()
    pBdr = OxmlElement("w:pBdr")
    bottom = OxmlElement("w:bottom")
    bottom.set(qn("w:val"), "single")
    bottom.set(qn("w:sz"), "6")
    bottom.set(qn("w:space"), "4")
    bottom.set(qn("w:color"), "4F46E5")
    pBdr.append(bottom)
    pPr.append(pBdr)

def _section_heading(doc, title):
    p = doc.add_paragraph()
    p.paragraph_format.space_before = Pt(12)
    p.paragraph_format.space_after  = Pt(4)
    run = p.add_run(title.upper())
    _set_font(run, size=10, bold=True, color=PRIMARY)
    _add_border_bottom(p)
    return p

def _bullet(doc, text):
    p = doc.add_paragraph(style="List Bullet")
    p.paragraph_format.space_before = Pt(1)
    p.paragraph_format.space_after  = Pt(1)
    p.paragraph_format.left_indent  = Cm(0.5)
    run = p.add_run(text)
    _set_font(run, size=9.5)
    return p

def generate_docx(data: dict) -> bytes:
    doc = Document()

    # Page setup — A4, narrow margins
    section = doc.sections[0]
    section.page_width  = Cm(21)
    section.page_height = Cm(29.7)
    section.top_margin    = Cm(1.5)
    section.bottom_margin = Cm(1.5)
    section.left_margin   = Cm(2)
    section.right_margin  = Cm(2)

    # Remove default paragraph spacing
    style = doc.styles["Normal"]
    style.font.name = "Arial"
    style.font.size = Pt(10)

    # ── Header: Name + title + contact ──────────────────────────────
    name_p = doc.add_paragraph()
    name_p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    name_run = name_p.add_run(data.get("nome", ""))
    _set_font(name_run, size=18, bold=True, color=PRIMARY)

    if titulo := data.get("titulo"):
        title_p = doc.add_paragraph()
        title_p.alignment = WD_ALIGN_PARAGRAPH.CENTER
        t_run = title_p.add_run(titulo)
        _set_font(t_run, size=11, color=MUTED)

    contato = data.get("contato", {})
    contact_parts = []
    for key in ("email", "telefone", "linkedin", "github", "localizacao"):
        val = contato.get(key)
        if val:
            contact_parts.append(val)
    if contact_parts:
        contact_p = doc.add_paragraph()
        contact_p.alignment = WD_ALIGN_PARAGRAPH.CENTER
        c_run = contact_p.add_run("  |  ".join(contact_parts))
        _set_font(c_run, size=9, color=MUTED)

    # ── Resumo Profissional ──────────────────────────────────────────
    if resumo := data.get("resumo_profissional"):
        _section_heading(doc, "Resumo Profissional")
        p = doc.add_paragraph()
        p.paragraph_format.space_after = Pt(4)
        run = p.add_run(resumo)
        _set_font(run, size=9.5)

    # ── Experiência ──────────────────────────────────────────────────
    experiencias = data.get("experiencia", [])
    if experiencias:
        _section_heading(doc, "Experiência Profissional")
        for exp in experiencias:
            p = doc.add_paragraph()
            p.paragraph_format.space_before = Pt(6)
            p.paragraph_format.space_after  = Pt(1)
            cargo_run = p.add_run(exp.get("cargo", ""))
            _set_font(cargo_run, size=10, bold=True, color=BLACK)
            p.add_run("  ")
            empresa_run = p.add_run(f"@ {exp.get('empresa', '')}")
            _set_font(empresa_run, size=10, color=MUTED)
            periodo_run = p.add_run(f"  •  {exp.get('periodo', '')}")
            _set_font(periodo_run, size=9, color=MUTED, italic=True)

            for resp in exp.get("responsabilidades", []):
                _bullet(doc, resp)

    # ── Projetos ─────────────────────────────────────────────────────
    projetos = data.get("projetos", [])
    if projetos:
        _section_heading(doc, "Projetos")
        for proj in projetos:
            p = doc.add_paragraph()
            p.paragraph_format.space_before = Pt(5)
            p.paragraph_format.space_after  = Pt(1)
            nome_run = p.add_run(proj.get("nome", ""))
            _set_font(nome_run, size=10, bold=True, color=PRIMARY)
            techs = proj.get("tecnologias", [])
            if techs:
                tech_run = p.add_run(f"  [{', '.join(techs)}]")
                _set_font(tech_run, size=9, color=MUTED)
            if desc := proj.get("descricao"):
                d_p = doc.add_paragraph()
                d_p.paragraph_format.left_indent = Cm(0.3)
                d_p.paragraph_format.space_after = Pt(1)
                d_run = d_p.add_run(desc)
                _set_font(d_run, size=9.5)
            if url := proj.get("url"):
                u_p = doc.add_paragraph()
                u_p.paragraph_format.left_indent = Cm(0.3)
                u_p.paragraph_format.space_after = Pt(1)
                u_run = u_p.add_run(url)
                _set_font(u_run, size=9, color=MUTED, italic=True)

    # ── Habilidades ──────────────────────────────────────────────────
    habilidades = data.get("habilidades", {})
    skill_lines = []
    if langs := habilidades.get("linguagens"):
        skill_lines.append(("Linguagens", ", ".join(langs)))
    if fws := habilidades.get("frameworks"):
        skill_lines.append(("Frameworks", ", ".join(fws)))
    if tools := habilidades.get("ferramentas"):
        skill_lines.append(("Ferramentas", ", ".join(tools)))
    if outros := habilidades.get("outros"):
        skill_lines.append(("Outros", ", ".join(outros)))

    if skill_lines:
        _section_heading(doc, "Habilidades")
        for label, value in skill_lines:
            p = doc.add_paragraph()
            p.paragraph_format.space_after = Pt(2)
            label_run = p.add_run(f"{label}: ")
            _set_font(label_run, size=9.5, bold=True)
            val_run = p.add_run(value)
            _set_font(val_run, size=9.5)

    # ── Formação ─────────────────────────────────────────────────────
    formacao = data.get("formacao", [])
    if formacao:
        _section_heading(doc, "Formação")
        for edu in formacao:
            p = doc.add_paragraph()
            p.paragraph_format.space_before = Pt(3)
            p.paragraph_format.space_after  = Pt(1)
            curso_run = p.add_run(edu.get("curso", ""))
            _set_font(curso_run, size=10, bold=True)
            inst_run = p.add_run(f"  —  {edu.get('instituicao', '')}")
            _set_font(inst_run, size=10, color=MUTED)
            if ano := edu.get("ano_conclusao"):
                ano_run = p.add_run(f"  ({ano})")
                _set_font(ano_run, size=9, color=MUTED, italic=True)

    # ── Idiomas ──────────────────────────────────────────────────────
    idiomas = data.get("idiomas", [])
    if idiomas:
        _section_heading(doc, "Idiomas")
        parts = [f"{i.get('idioma', '')} ({i.get('nivel', '')})" for i in idiomas]
        p = doc.add_paragraph()
        run = p.add_run("  •  ".join(parts))
        _set_font(run, size=9.5)

    # ── Certificações ────────────────────────────────────────────────
    certs = data.get("certificacoes", [])
    if certs:
        _section_heading(doc, "Certificações")
        for cert in certs:
            _bullet(doc, cert)

    buf = io.BytesIO()
    doc.save(buf)
    return buf.getvalue()

File: src/adapters/DocxAdapter.js
/**
 * DocxAdapter — gera um arquivo .docx de curriculo a partir do JSON reconstruido.
 * Usa a biblioteca docx (funciona no browser via ESM).
 */
import {
  Document, Packer, Paragraph, TextRun, Table, TableRow, TableCell,
  AlignmentType, LevelFormat, WidthType, BorderStyle, ShadingType,
  ExternalHyperlink, TabStopType, TabStopPosition, HeadingLevel,
} from 'docx'

// ── Cores e estilos ────────────────────────────────────────────────
const PRIMARY = '4F46E5'   // indigo
const MUTED   = '64748B'
const BLACK   = '1E293B'
const BORDER  = { style: BorderStyle.NONE, size: 0, color: 'FFFFFF' }
const NO_BORDERS = { top: BORDER, bottom: BORDER, left: BORDER, right: BORDER }

function run(text, opts = {}) {
  return new TextRun({ text: String(text || ''), font: 'Arial', color: BLACK, ...opts })
}

function muted(text, opts = {}) {
  return new TextRun({ text: String(text || ''), font: 'Arial', color: MUTED, size: 20, ...opts })
}

function sectionHeading(title) {
  return new Paragraph({
    children: [new TextRun({ text: title.toUpperCase(), bold: true, font: 'Arial', color: PRIMARY, size: 22 })],
    spacing: { before: 280, after: 80 },
    border: { bottom: { style: BorderStyle.SINGLE, size: 8, color: PRIMARY, space: 4 } },
  })
}

function bullet(text) {
  return new Paragraph({
    numbering: { reference: 'bullets', level: 0 },
    children: [run(text, { size: 22 })],
    spacing: { after: 40 },
  })
}

function spacer(size = 120) {
  return new Paragraph({ children: [new TextRun('')], spacing: { after: size } })
}

// ── Montagem do documento ──────────────────────────────────────────
export async function generate(data) {
  const children = []

  // ── Nome e titulo ──────────────────────────────────────────────
  children.push(new Paragraph({
    children: [new TextRun({ text: data.nome || 'Nome', bold: true, font: 'Arial', size: 52, color: BLACK })],
    alignment: AlignmentType.CENTER,
    spacing: { after: 80 },
  }))

  if (data.titulo_profissional) {
    children.push(new Paragraph({
      children: [new TextRun({ text: data.titulo_profissional, font: 'Arial', size: 24, color: MUTED })],
      alignment: AlignmentType.CENTER,
      spacing: { after: 100 },
    }))
  }

  // ── Contato ────────────────────────────────────────────────────
  const c = data.contato || {}
  const contatoItens = [c.email, c.telefone, c.linkedin, c.github, c.portfolio].filter(Boolean)
  if (contatoItens.length > 0) {
    children.push(new Paragraph({
      children: contatoItens.flatMap((item, i) => [
        new TextRun({ text: item, font: 'Arial', size: 20, color: MUTED }),
        ...(i < contatoItens.length - 1 ? [new TextRun({ text: '  |  ', font: 'Arial', size: 20, color: MUTED })] : []),
      ]),
      alignment: AlignmentType.CENTER,
      spacing: { after: 200 },
    }))
  }

  // ── Resumo profissional ────────────────────────────────────────
  if (data.resumo_profissional) {
    children.push(sectionHeading('Resumo Profissional'))
    children.push(new Paragraph({
      children: [run(data.resumo_profissional, { size: 22 })],
      spacing: { after: 60 },
    }))
  }

  // ── Experiencia ────────────────────────────────────────────────
  if (data.experiencias?.length > 0) {
    children.push(sectionHeading('Experiência Profissional'))
    for (const exp of data.experiencias) {
      children.push(new Paragraph({
        children: [
          new TextRun({ text: exp.cargo || '', bold: true, font: 'Arial', size: 24, color: BLACK }),
          new TextRun({ text: '\t' + (exp.periodo || ''), font: 'Arial', size: 20, color: MUTED }),
        ],
        tabStops: [{ type: TabStopType.RIGHT, position: TabStopPosition.MAX }],
        spacing: { before: 120, after: 40 },
      }))
      children.push(new Paragraph({
        children: [new TextRun({ text: exp.empresa || '', font: 'Arial', size: 22, color: PRIMARY })],
        spacing: { after: 60 },
      }))
      for (const desc of (exp.descricoes || [])) {
        children.push(bullet(desc))
      }
      children.push(spacer(60))
    }
  }

  // ── Projetos ───────────────────────────────────────────────────
  if (data.projetos?.length > 0) {
    children.push(sectionHeading('Projetos'))
    for (const proj of data.projetos) {
      const linkRun = proj.url
        ? new ExternalHyperlink({
            link: proj.url,
            children: [new TextRun({ text: proj.url, font: 'Arial', size: 20, color: PRIMARY, underline: {} })],
          })
        : null
      children.push(new Paragraph({
        children: [
          new TextRun({ text: proj.nome || '', bold: true, font: 'Arial', size: 23, color: BLACK }),
          new TextRun({ text: '  —  ' + (proj.tecnologias || ''), font: 'Arial', size: 20, color: MUTED }),
        ],
        spacing: { before: 100, after: 40 },
      }))
      children.push(new Paragraph({
        children: [run(proj.descricao || '', { size: 22 })],
        spacing: { after: 40 },
      }))
      if (linkRun) {
        children.push(new Paragraph({ children: [linkRun], spacing: { after: 80 } }))
      }
    }
  }

  // ── Habilidades ────────────────────────────────────────────────
  const hab = data.habilidades || {}
  const habItems = [
    hab.linguagens?.length ? { label: 'Linguagens', value: hab.linguagens.join(', ') } : null,
    hab.frameworks?.length ? { label: 'Frameworks & Libs', value: hab.frameworks.join(', ') } : null,
    hab.ferramentas?.length ? { label: 'Ferramentas', value: hab.ferramentas.join(', ') } : null,
  ].filter(Boolean)

  if (habItems.length > 0) {
    children.push(sectionHeading('Habilidades Técnicas'))
    for (const item of habItems) {
      children.push(new Paragraph({
        children: [
          new TextRun({ text: item.label + ': ', bold: true, font: 'Arial', size: 22, color: BLACK }),
          run(item.value, { size: 22 }),
        ],
        spacing: { after: 60 },
      }))
    }
  }

  // ── Formacao ───────────────────────────────────────────────────
  if (data.formacao?.length > 0) {
    children.push(sectionHeading('Formação'))
    for (const form of data.formacao) {
      children.push(new Paragraph({
        children: [
          new TextRun({ text: form.curso || '', bold: true, font: 'Arial', size: 22, color: BLACK }),
          new TextRun({ text: '\t' + (form.periodo || ''), font: 'Arial', size: 20, color: MUTED }),
        ],
        tabStops: [{ type: TabStopType.RIGHT, position: TabStopPosition.MAX }],
        spacing: { before: 80, after: 30 },
      }))
      children.push(new Paragraph({
        children: [new TextRun({ text: form.instituicao || '', font: 'Arial', size: 21, color: MUTED })],
        spacing: { after: 80 },
      }))
    }
  }

  // ── Idiomas e Certificacoes ────────────────────────────────────
  const hasIdiomas = data.idiomas?.length > 0
  const hasCerts = data.certificacoes?.length > 0

  if (hasIdiomas) {
    children.push(sectionHeading('Idiomas'))
    for (const id of data.idiomas) {
      children.push(new Paragraph({
        children: [
          new TextRun({ text: (id.idioma || '') + ': ', bold: true, font: 'Arial', size: 22, color: BLACK }),
          run(id.nivel || '', { size: 22 }),
        ],
        spacing: { after: 50 },
      }))
    }
  }

  if (hasCerts) {
    children.push(sectionHeading('Certificações'))
    for (const cert of data.certificacoes) {
      children.push(bullet(cert))
    }
  }

  const doc = new Document({
    numbering: {
      config: [{
        reference: 'bullets',
        levels: [{
          level: 0,
          format: LevelFormat.BULLET,
          text: '•',
          alignment: AlignmentType.LEFT,
          style: { paragraph: { indent: { left: 440, hanging: 220 } } },
        }],
      }],
    },
    styles: {
      default: { document: { run: { font: 'Arial', size: 22, color: BLACK } } },
    },
    sections: [{
      properties: {
        page: {
          size: { width: 11906, height: 16838 }, // A4
          margin: { top: 1000, right: 1000, bottom: 1000, left: 1000 },
        },
      },
      children,
    }],
  })

  return Packer.toBlob(doc)
}
